### PR TITLE
LPD-89355 Preserve folder membership on no-op upsert

### PR DESCRIPTION
**Pull request by adolfopa**
https://liferay.atlassian.net/browse/LPD-89355

## What is being fixed

On a 2025.Q3 → 2026.Q1 upgrade, every cluster pod fails to boot because the CMS site initializer can no longer re-import its folders. `01-object-folder.batch-engine-data.json` ships with `"objectFolderItems": []` for every CMS folder. On a fresh install those folders are empty, so the upsert is a no-op; on upgrade the existing CMS object definitions are evicted to the default folder, which LPD-83357's `ObjectDefinitionModelListener` rejects with `ObjectDefinitionObjectFolderIdException`.

## How it is being fixed

`ObjectFolderResourceImpl._addObjectFolderResources` now skips the eviction loop when the upsert payload supplies no `objectFolderItems`, so a metadata-only refresh leaves existing membership untouched. The default-folder lookup that the loop needs has been swapped over to the dedicated `ObjectFolderLocalService.getDefaultObjectFolder` helper. As a defensive cleanup on the same path, `ObjectDefinitionLocalServiceImpl.updateObjectFolderId` now short-circuits when the resolved new folder id matches the current one, with an even earlier short-circuit on the raw input that avoids an extra folder lookup.

A regression test in `ObjectFolderResourceTest.testPutObjectFolder` re-`PUT`s a folder containing a definition with empty `objectFolderItems` and asserts the definition stays in place.

### CODE REVIEW

**Comment by marcelabc** [comment]
Hey @nicolas-sss 
Could you review this PR, please? :smile:

File: modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectFolderResourceImpl.java [hunk 1]
```diff
@@ -306,27 +305,28 @@ private void _addObjectFolderResources(
...
 
-		com.liferay.object.model.ObjectFolder
-			defaultServiceBuilderObjectFolder =
-				_objectFolderService.getObjectFolderByExternalReferenceCode(
-					ObjectFolderConstants.EXTERNAL_REFERENCE_CODE_DEFAULT,
-					contextCompany.getCompanyId());
+		if (!objectFolderItems.isEmpty()) {
+			com.liferay.object.model.ObjectFolder
+				defaultServiceBuilderObjectFolder =
+					_objectFolderLocalService.getDefaultObjectFolder(
```
Comment (nicolas-sss) [inline modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectFolderResourceImpl.java:311]: @adolfopa Swap from `_objectFolderService` to `_objectFolderLocalService` drops the permission check on the default-folder fetch. Matches the pattern at line 193, probably fine for internal eviction. Deliberate, or fallout from picking the dedicated helper?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java [hunk 2]
```diff
@@ -1328,9 +1328,18 @@ public ObjectDefinition updateObjectFolderId(
...
+		if (objectFolderId == oldObjectFolderId) {
+			return objectDefinition;
+		}
+
+		long newObjectFolderId = _getObjectFolderId(
+			objectDefinition.getCompanyId(), objectFolderId);
+
+		if (newObjectFolderId == oldObjectFolderId) {
+			return objectDefinition;
+		}
```
Comment (nicolas-sss) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java:1340]: @adolfopa No-op moves now skip persistence.update entirely, so no `ObjectDefinition` model listener fires. Right call for `LPD-83357`. 
Any other listeners registered in ObjectDefinition that perform work in the update() method (audit, indexing, custom hooks) wouldn't be also skipped? expected behavior? 
If not, the check could be on the listener side instead?

**Review by nicolas-sss — APPROVED** [review]
Hey @adolfopa Looks legit from our side, just some questions, please take a look. If any of them make sense, feel free to forward directly after fixing.

The ticket lists `isLinkedToObjectFolder` returning false-when-primary as a compounding factor, but the PR does not touch it.

An admin PUT on CMS folder `F` with `objectFolderItems` that omit a CMS definition `D` whose primary is `F` would still fail through `isLinkedToObjectFolder` returning false.

Edge case not worth treating?

**Comment by nicolas-sss** [comment]
ci:test:sf

File: modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectFolderResourceImpl.java (hunk 1 above)
Comment (adolfopa) [reply modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectFolderResourceImpl.java:311]: @nicolas-sss It is deliberate: there's no way to get an object folder by name from the non-local service (would require api changes), all other calls in that method are also local (consistency) and permissions are already checked from calling sites (should be safe).

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java (hunk 2 above)
Comment (adolfopa) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java:1340]: This is expected. There's no actual update, so why invoke the model listeners? Listeners should only fire when there's an actual persistence level change; things that depend on service methods being invoked should be implemented as service wrappers, not model listeners.

**Comment by adolfopa** [comment]
@nicolas-sss Thanks for your comments. See my answers and let me know if you want me to address those things anyway.

About `isLinkedToObjectFolder`, the implementation looks wrong, but I didn't change it because its been like that for a long time, and inverting the condition now should break a ton of things. I think that's something that will need to be done carefully in a separate ticket.

**Comment by nicolas-sss** [comment]
@adolfopa Thanks for the answers, all make sense to me. I'll create the follow-up ticket for `isLinkedToObjectFolder` and link it here soon. Going to forward this myself since it has been parked for a while. Thanks again for the help!

**Comment by nicolas-sss** [comment]
ci:forward
